### PR TITLE
Python script for IsolationForest Outlier Detection

### PR DESCRIPTION
**Pull request by esther11**

### CODE REVIEW

File: outlier_detection/README.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+For running outlier_detection_ilf.py:
+
+Needed packages: pandas, math, sklearn.
+
+Data Preparation: store data in a csv file, including the dimensions (must be numeric values) 
+that are going to be used for identifing outliers. The first line of csv stores dimension names.
```
Comment (lsloan) [inline outlier_detection/README.md]: Please add details about the data preparation:

* How is the data transformed to CSV from JSON?  I didn't know this program used a CSV as input, not the JSON.  Do you use Tableau for that transform?  Could this program do the transform or does Tableau do something complex?  (For example, src/analyze/events.py uses JSON directly.)

* Could you add an example input CSV file?  That would let us test the program without requiring Tableau.

File: outlier_detection/outlier_detect_ilf.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
+## Python verson 3.6
```
Comment (lsloan) [inline outlier_detection/outlier_detect_ilf.py]: Add code to check the version of Python being used.  Display an error and exit if the version doesn't meet minimum requirements.  (Hint: See `sys.version_info`)

File: outlier_detection/outlier_detect_ilf.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+## Replace the items in <> before running
```
Comment (lsloan) [inline outlier_detection/outlier_detect_ilf.py]: Add support for command-line options and arguments rather than requiring the program be edited before running.  For examples, see `extract.py` in this repo.

File: outlier_detection/outlier_detect_ilf.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+
+import pandas as pd
```
Comment (lsloan) [inline outlier_detection/outlier_detect_ilf.py]: Organize all the imports at the beginning of the program.

File: outlier_detection/outlier_detect_ilf.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+# ------------------------------------------------------------
+from math import log
+
+def cal_log(x):
+    if x == 0:
+        return 0
+    else:
+        return log(x)
+
+data['<column_name>'] = list(map(cal_log, data['<column_name>']))
```
Comment (lsloan) [inline outlier_detection/outlier_detect_ilf.py]: Move this into a block at the end of the file, which will be the main block of the program.  Rough organization of the program overall:

1. `import` statements
1. `def` of one or more functions
1. main program block

**Review by lsloan — CHANGES_REQUESTED** [review]
See my various comments about specific parts of the code.  Also, the code in this repo has been reorganized into a `src` directory.  Bring that new organization into your copy of the project, then put your program in a `src/analyze/tableau` directory.

File: src/analyze/tableau/outlier_detection/outlier_detect_ilf.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    type=bool,
+    dest='use_log',
+    default=True,
+    help='''Whether or not use log value of Attempt Duration.
+            Default: %(default)s'''
+)
+args = argParser.parse_args()
+
+
+def __main__(f_name=None):
```
Comment (lsloan) [inline src/analyze/tableau/outlier_detection/outlier_detect_ilf.py]: Although the Python internal symbol for the name of the program is `'__main__'`, you don't need to name your function `__main__`, too.  In fact, it might be bad to do so, since symbol names that begin and end with `__` usually have special purposes.  (Either now or in the future.)

Name your main function simply `main` instead.

File: src/analyze/tableau/outlier_detection/outlier_detect_ilf.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    if f_name == None or f_name == []:
```
Comment (lsloan) [inline src/analyze/tableau/outlier_detection/outlier_detect_ilf.py]: I think you need to test `f_name` for the empty string (`''`), not an empty list (`[]`).

File: src/analyze/tableau/outlier_detection/outlier_detect_ilf.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+def ilf_predict(data, contamination, use_log):
+    if use_log:
+        use_log_value(data)
+    ilf = IsolationForest(contamination=contamination) # Consider 5% of the data as "outliers" if not specified
+    use_df = data[data.columns[1:]]
+    ilf.fit(use_df) # Fit the model
+    data['IsInlier'] = ilf.predict(use_df) # Unsupervised learning and predicting outliers
+
+
+argParser = argparse.ArgumentParser(
```
Comment (lsloan) [inline src/analyze/tableau/outlier_detection/outlier_detect_ilf.py]: Lines 55-80 should be inside the `main` function.  Whenever possible, code should be inside a function.

File: src/analyze/tableau/outlier_detection/outlier_detect_ilf.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
+import sys
+if sys.version_info < (3, 6):
```
Comment (lsloan) [inline src/analyze/tableau/outlier_detection/outlier_detect_ilf.py]: This is a good test for Python 3.6, but I think it should be in the `main` function.  It's tempting to keep it here, because it's a convenient place to see the version requirement quickly.  However, it goes against the usual "everything in functions" practice.

**Review by lsloan — COMMENTED** [review]
I may have a couple more comments later, but I thought I'd give you some comments now so you will have something to work with.

File: src/analyze/tableau/outlier_detection/README.md [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+	AttemptDuration: length of time PR user spent on the problem (measured in seconds).
+
+Command-line options:
+	
+	--perc: Determine percentage of events in data to be considered "outliers". Default value 0.05.
+	--uselog: Decide whether or not to use log value of AttemptDuration. Default value True.
+
+Sample command:
+	
+	python3 --perc 0.1 --uselog f outlier_detection_ilf.py prdata.json
```
Comment (lsloan) [inline src/analyze/tableau/outlier_detection/README.md]: Some questions about the `--uselog` option:

1. Did you run the program using `--uselog f`?
  I think it won't work as expected, because `argparse` doesn't parse "f" as `False`.  If the argument to a Boolean option contains **_anything_**, then it means `True`.

1. Is the user required to always specify this option?

1. Will the option work if it's not given any "t" or "f" argument?

I recommend treating this option as a Boolean flag instead, without an argument.  Because it's called "uselog", I expect if it's given, then the `log()` function will be used.  If it's not given, then the default behavior would be to not use `log()`.  That will avoid the problem of parsing a Boolean argument.

In the program, use:

```python
parser.add_argument('--uselog', dest='use_log', action='store_true', default=False, help='Use log value of Attempt Duration.')
```

Note that `type` can't be used with `action='store_true'` and `help` is shortened and doesn't show the default value.

File: src/analyze/tableau/outlier_detection/outlier_detect_ilf.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+    return pd.DataFrame(events, columns=['OpenlrsSourceId', 'IsComplete', 'IsCorrect', 'AttemptCount', 'AttemptDuration'])
+
+def cal_log(x):
+    if x == 0:
+        return 0
+    else:
+        return log(x)
+
+def use_log_value(data):
+    data['AttemptDuration'] = list(map(cal_log, data['AttemptDuration']))
```
Comment (lsloan) [inline src/analyze/tableau/outlier_detection/outlier_detect_ilf.py]: I haven't tested this idea, but I think the program will be faster if you use a lambda in place of `cal_log`:

```python
    data['AttemptDuration'] = map(lambda x: 0 if x == 0 else log(x), data['AttemptDuration'])
```

This faster processing may be important when used with a large amount of data.  Also note that `list()` is unnecessary here.

**Review by lsloan — CHANGES_REQUESTED** [review]
It turns out that I had more than a couple more comments.  Note the high priority comments are about the bug with `--uselog` and replacing the output file.

After you've received comments from @pushyamig and @zqian, make the necessary changes, then let's go over the code together.

**Comment by esther11** [comment]
Moved everything into functions; modified argparse positional argument "file" handling.

File: src/analyze/tableau/outlier_detection/outlier_detect_ilf.py [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+    args = argParser.parse_args()
+
+    if args.file == '':
+        print("No file name provided.")
+        sys.exit(1)
+
+    try:
+        data = read_file(args.file)
+        ilf_predict(data, args.contamination, args.use_log)
+        data.to_csv('outlier_detect.csv', columns=['OpenlrsSourceId', 'IsInlier'], index=False)
```
Comment (lsloan) [inline src/analyze/tableau/outlier_detection/outlier_detect_ilf.py]: Always writing to a file named `outlier_detect.csv` is a problem.  Since the default write mode is used, if that file already exists, it will be replaced with a new one.  If a user uses this program for multiple visualizations, it will be confusing or frustrating if they didn't expect the file to be replaced.  Try one or more of these solutions:

* Let the user to specify input and output filenames.  E.g., it could be run as: `python3 --perc 0.1 --uselog f outlier_detection_ilf.py prdata.json` **_`output_filename.csv`_**

  However, this requires displaying an error message if the user didn't specify an output filename.

* Send the output to stdout and let the user redirect it to a file.  E.g., `python3 --perc 0.1 --uselog f outlier_detection_ilf.py prdata.json` **_`> output_filename.csv`_**
* A combination of the above two.

  If the user specified an output filename, use it.  If no output filename was specified, send the output to stdout.  One benefit of this is that no error message needs to be displayed if the user didn't specify an output filename.

File: src/analyze/tableau/outlier_detection/outlier_detect_ilf.py [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+    except:
+        print("Sorry, cannot read this file.")
```
Comment (lsloan) [inline src/analyze/tableau/outlier_detection/outlier_detect_ilf.py]: When it can't read the file, it looks like the `'Sorry, cannot read this file.'` **_AND_** `'Please check your working directory for output file "outlier_detect.csv".'` messages will both be shown.  Because after the "cannot read" message is shown, processing will continue with the next line.  This is a minor bug, but it could confuse users.

Also, the "cannot read" message is too specific for the `except` keyword without giving an exception name.  That is, the call to `read_file` may cause an error, but `ilf_predict` and `to_csv` could, too.  It might be simpler to catch the exception, give a general message, like "An error has occurred:", followed by the text of the exception.

**Review by lsloan — CHANGES_REQUESTED** [review]
Some of the code was moved before I sent my last set of comments, so a couple of them became "outdated".  I've added the comments again, in their new locations.

File: src/analyze/tableau/outlier_detection/README.md [hunk 14]
```diff
@@ -0,0 +1,26 @@
+Program outlier_detection_ilf.py process Caliper JSON format data collected from Problem Roulette (PR). Refer to https://github.com/tl-its-umich-edu/caliper-event-analysis/tree/master/src/extract/ for scripts that extract data from Problem Roulette's database.
+
+The program should be excecuted from command line. You will need Python version 3.6+ and packages [pandas, math, sklearn, argparse, json] to run the program.
+
+The program use Isolation Forest algorithm to detect rare ones among all PR problem-dealing events and consider them "outliers". Dimensions used in the to detect outliers include:
```
Comment (pushyamig) [inline src/analyze/tableau/outlier_detection/README.md]: I think a link needs to be provided from where the reading source for the Isolation Forest algorithm, so people who are unaware can follow think and read vs searching on there own

File: src/analyze/tableau/outlier_detection/outlier_detect_ilf.py [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+def cal_log(x):
+    if x == 0:
+        return 0
+    else:
+        return log(x)
+
+def use_log_value(data):
+    data['AttemptDuration'] = list(map(cal_log, data['AttemptDuration']))
+
+def ilf_predict(data, contamination, use_log):
```
Comment (pushyamig) [inline src/analyze/tableau/outlier_detection/outlier_detect_ilf.py]: A better name than `ilf` can be provided here

File: src/analyze/tableau/outlier_detection/outlier_detect_ilf.py [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+    ilf = IsolationForest(contamination=contamination) # Consider 5% of the data as "outliers" if not specified
+    use_df = data[data.columns[1:]]
+    ilf.fit(use_df) # Fit the model
+    data['IsInlier'] = ilf.predict(use_df) # Unsupervised learning and predicting outliers
+
+
+def main():
+
+    if sys.version_info < (3, 6):
+        print('Sorry, need Python version 3.6+')
```
Comment (pushyamig) [inline src/analyze/tableau/outlier_detection/outlier_detect_ilf.py]: A better statement would be `Python version 3.6+ is required`

File: src/analyze/tableau/outlier_detection/outlier_detect_ilf.py [hunk 17]
```diff
@@ -0,0 +1,101 @@
+import sys
+import argparse
+import json
+import pandas as pd
+from math import log
+from sklearn.ensemble import IsolationForest
+
+
+def read_file(f_name):
```
Comment (pushyamig) [inline src/analyze/tableau/outlier_detection/outlier_detect_ilf.py]: Some comments needs to be added to this function what the purpose/outcome from this call, the function name is too generic and this call is doing little more that reading a file
